### PR TITLE
applications: serial_lte_modem: Code restructure

### PR DESCRIPTION
**Pull request by junqingzou**
Restruct the project like below:
- slm_at_host.c: handle modem AT commands, serial interface, command
and data modes, etc.
- slm_at_commands.c(new): handle all SLM proprietary AT commands
- single call of AT Parser function for SLM proprietary AT command

Signed-off-by: Jun Qing Zou <jun.qing.zou@nordicsemi.no>

### CODE REVIEW

File: applications/serial_lte_modem/src/slm_at_commands.c [hunk 1]
```diff
@@ -0,0 +1,580 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
```
Comment (rlubos) [inline applications/serial_lte_modem/src/slm_at_commands.c:4]: Please use the new license identifier: `LicenseRef-Nordic-5-Clause`

File: applications/serial_lte_modem/src/slm_at_commands.c [hunk 2]
```diff
@@ -0,0 +1,580 @@
...
+void enter_idle(void);
+void enter_sleep(bool wake_up);
+int set_uart_baudrate(uint32_t baudrate);
+int get_uart_baudrate(void);
+void rsp_send(const uint8_t *str, size_t len);
+
+#define SLM_VERSION	"\r\n#XSLMVER: \"1.6\"\r\n"
+
+/**@brief handle AT#XSLMVER commands
+ *  #XSLMVER
```
Comment (rlubos) [inline applications/serial_lte_modem/src/slm_at_commands.c:86]: Shouldn't this be `AT#SLMVER`?

File: applications/serial_lte_modem/src/slm_at_commands.c [hunk 3]
```diff
@@ -0,0 +1,580 @@
...
+	if (type == AT_CMD_TYPE_SET_COMMAND) {
+		rsp_send(SLM_VERSION, sizeof(SLM_VERSION) - 1);
+		ret = 0;
+	}
+
+	return ret;
+}
+
+/**@brief handle AT#XSLEEP commands
+ *  #XSLEEP[=<shutdown_mode>]
```
Comment (rlubos) [inline applications/serial_lte_modem/src/slm_at_commands.c:103]: `AT#XSLEEP`?

File: applications/serial_lte_modem/src/slm_at_commands.c [hunk 4]
```diff
@@ -0,0 +1,580 @@
...
+			LOG_ERR("AT parameter error");
+			ret = -EINVAL;
+		}
+	}
+
+	return ret;
+}
+
+/**@brief handle AT#XRESET commands
+ *  #XRESET
```
Comment (rlubos) [inline applications/serial_lte_modem/src/slm_at_commands.c:140]: `AT#XRESET`?

File: applications/serial_lte_modem/src/slm_at_commands.c [hunk 5]
```diff
@@ -0,0 +1,580 @@
...
+/**@brief AT command handler type. */
+typedef int (*slm_at_handler_t) (enum at_cmd_type);
+
+/**@brief AT command list item type. */
+typedef struct slm_at_cmd {
+	char *string;
+	slm_at_handler_t handler;
+} slm_at_cmd_t;
+
+static struct slm_work_t {
```
Comment (rlubos) [inline applications/serial_lte_modem/src/slm_at_commands.c:61]: No need for `_t` suffix, you don't use typedef here.

File: applications/serial_lte_modem/src/slm_at_commands.c [hunk 6]
```diff
@@ -0,0 +1,580 @@
...
+			datamode_size_limit = size_limit;
+			datamode_time_limit = time_limit;
+		} else {
+			return -EINVAL;
+		}
+		break;
+
+	case AT_CMD_TYPE_READ_COMMAND:
+		sprintf(rsp_buf, "#XDATACTRL: %d,%d\r\n",
+				datamode_size_limit, datamode_time_limit);
```
Comment (rlubos) [inline applications/serial_lte_modem/src/slm_at_commands.c:258]: Alignment

File: applications/serial_lte_modem/src/slm_at_commands.c [hunk 7]
```diff
@@ -0,0 +1,580 @@
...
+
+#if defined(CONFIG_SLM_MQTTC)
+int handle_at_mqtt_connect(enum at_cmd_type cmd_type);
+int handle_at_mqtt_publish(enum at_cmd_type cmd_type);
+int handle_at_mqtt_subscribe(enum at_cmd_type cmd_type);
+int handle_at_mqtt_unsubscribe(enum at_cmd_type cmd_type);
+#endif
+
+#if defined(CONFIG_SLM_HTTPC)
+int handle_AT_HTTPC_CONNECT(enum at_cmd_type cmd_type);
```
Comment (rlubos) [inline applications/serial_lte_modem/src/slm_at_commands.c:334]: Would be good to unify handlers style at some future cleanup (i. e. use lowercase).

File: applications/serial_lte_modem/src/slm_at_commands.c [hunk 8]
```diff
@@ -0,0 +1,580 @@
...
+	{"AT#XHTTPCREQ", handle_AT_HTTPC_REQUEST},
+#endif
+};
+
+int handle_at_clac(enum at_cmd_type cmd_type)
+{
+	int ret = -EINVAL;
+
+	if (cmd_type == AT_CMD_TYPE_SET_COMMAND) {
+		int total = sizeof(slm_at_cmd_list)/sizeof(slm_at_cmd_t);
```
Comment (rlubos) [inline applications/serial_lte_modem/src/slm_at_commands.c:410]: You could use `ARRAY_SIZE(slm_at_cmd_list)` utility macro to obtain number of array elements.

File: applications/serial_lte_modem/src/slm_at_commands.c [hunk 9]
```diff
@@ -0,0 +1,580 @@
...
+		ret = 0;
+	}
+
+	return ret;
+}
+
+int slm_at_parse(const char *at_cmd)
+{
+	int ret = -ENOENT;
+	int total = sizeof(slm_at_cmd_list)/sizeof(slm_at_cmd_t);
```
Comment (rlubos) [inline applications/serial_lte_modem/src/slm_at_commands.c:425]: Same here (`ARRAY_SIZE`).

File: applications/serial_lte_modem/src/slm_at_commands.c [hunk 10]
```diff
@@ -0,0 +1,580 @@
...
+
+	for (int i = 0; i < total; i++) {
+		if (slm_util_cmd_casecmp(at_cmd, slm_at_cmd_list[i].string)) {
+			enum at_cmd_type type = at_parser_cmd_type_get(at_cmd);
+
+			ret = at_parser_params_from_str(at_cmd, NULL,
+						&at_param_list);
```
Comment (rlubos) [inline applications/serial_lte_modem/src/slm_at_commands.c:432]: Alignment

**Review by rlubos — COMMENTED** [review]
Looks good overall, just some nits. Good to see the overall LOC reduction (less code to maintain).

File: applications/serial_lte_modem/src/slm_at_commands.c (hunk 2 above)
Comment (junqingzou) [reply applications/serial_lte_modem/src/slm_at_commands.c:103]: AT#X... for all SLM commands, as in the doc

File: applications/serial_lte_modem/src/slm_at_commands.c (hunk 5 above)
Comment (junqingzou) [reply applications/serial_lte_modem/src/slm_at_commands.c:62]: updated

File: applications/serial_lte_modem/src/slm_at_commands.c (hunk 7 above)
Comment (junqingzou) [reply applications/serial_lte_modem/src/slm_at_commands.c:365]: updated

File: applications/serial_lte_modem/src/slm_at_commands.c (hunk 8 above)
Comment (junqingzou) [reply applications/serial_lte_modem/src/slm_at_commands.c]: Taken with gratitude

File: applications/serial_lte_modem/src/slm_at_commands.c (hunk 9 above)
Comment (junqingzou) [reply applications/serial_lte_modem/src/slm_at_commands.c]: Taken with gratitude

File: applications/serial_lte_modem/src/slm_at_tcpip.c [hunk 11]
```diff
@@ -685,7 +641,7 @@ static int do_recvfrom(uint16_t length)
 		ret = slm_util_htoa(rx_data, ret, data_hex, size);
 		if (ret > 0) {
 			rsp_send(data_hex, ret);
-			sprintf(rsp_buf, "\r\n#XRECVFROM: %d,%d\r\n",
+			sprintf(rsp_buf, "\r\n\r\n#XRECVFROM: %d,%d\r\n",
 				DATATYPE_HEXADECIMAL, ret);
```
Comment (lats1980) [inline applications/serial_lte_modem/src/slm_at_tcpip.c:645]: repeated CRLF?

File: applications/serial_lte_modem/src/slm_at_tcpip.c [hunk 12]
```diff
@@ -694,7 +650,7 @@ static int do_recvfrom(uint16_t length)
 		}
 	} else {
 		rsp_send(rx_data, ret);
-		sprintf(rsp_buf, "\r\n#XRECVFROM: %d,%d\r\n",
+		sprintf(rsp_buf, "\r\n\r\n#XRECVFROM: %d,%d\r\n",
 			DATATYPE_PLAINTEXT, ret);
```
Comment (lats1980) [inline applications/serial_lte_modem/src/slm_at_tcpip.c:654]: repeated CRLF?

**Review by lats1980 — COMMENTED** [review]
Looks good, except some fix for URC format seems wrong (repeated CRLF before URC). Can you clarify?

**Comment by junqingzou** [comment]
> Looks good, except some fix for URC format seems wrong (repeated CRLF before URC). Can you clarify?

Side product of Change All. Will correct

File: applications/serial_lte_modem/src/slm_at_commands.c (hunk 7 above)
Comment (junqingzou) [reply applications/serial_lte_modem/src/slm_at_commands.c]: Fixed

**Pull request merged**